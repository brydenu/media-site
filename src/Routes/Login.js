import React, { useContext, useState } from "react";
import { useNavigate } from "react-router";
import AppContext from "../Helpers/Context";
import Form from "../Common/Form";
import Card from "../Common/Card";
import Backend from "../api";

/**
 * Login page.
 *
 * On submit, sends request using form data to backend, to attempt to login.
 *
 * On successful login, redirects to home page.
 */
export default function Login() {
    const { setToken } = useContext(AppContext).tokenState;
    const { setUser } = useContext(AppContext).userState;
    const navigate = useNavigate();
    const [userInput, setUserInput] = useState({
        username: "",
        password: "",
    });
    const headerContent = <h1 className="login-title">Login</h1>;

    const formFields = [
        {
            label: "Username",
            type: "text",
            fieldName: "username",
            required: true,
        },
        {
            label: "Password",
            type: "password",
            fieldName: "password",
            required: true,
        },
    ];

    const handleSubmit = async (evt) => {
        evt.preventDefault();
        const res = await Backend.loginUser(userInput);
        const { queries } = await Backend.getQueryHistory(res.user.username);
        res.user.queries = queries;
        setToken(res.token);
        setUser(res.user);
        navigate("/");
    };

    const bodyContent = (
        <Form
            fields={formFields}
            formClass="login"
            handleSubmit={handleSubmit}
            inputState={[userInput, setUserInput]}
            buttonLabel="Login"
        />
    );
    return (
        <div className="main-login">
            <Card header={headerContent} body={bodyContent} cardClass="login" />
        </div>
    );
}
