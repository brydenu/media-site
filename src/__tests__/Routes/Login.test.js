import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import Login from "../../Routes/Login";
import TestContext from "../../contextForTests";

it("renders without crashing", function () {
    render(
        <MemoryRouter>
            <TestContext>
                <Login />
            </TestContext>
        </MemoryRouter>
    );
});

it("matches snapshot", function () {
    const { asFragment } = render(
        <MemoryRouter>
            <TestContext>
                <Login />
            </TestContext>
        </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});
