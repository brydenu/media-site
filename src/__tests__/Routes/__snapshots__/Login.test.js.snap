// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="main-login"
  >
    <div
      class="card-wrapper login"
    >
      <div
        class="card-content login"
      >
        <div
          class="card-header "
        >
          <h1
            class="login-title"
          >
            Login
          </h1>
        </div>
        <div
          class="card-body login"
        >
          <form
            class="custom-form login"
          >
            <div
              class="form-group username"
            >
              <label
                for="username"
              >
                Username
              </label>
              <input
                autocomplete="off"
                name="username"
                placeholder=""
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group password"
            >
              <label
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="off"
                name="password"
                placeholder=""
                type="password"
                value=""
              />
            </div>
            <div
              class="clickables-group"
            >
              <div
                class="checkbox-group"
              >
                <div
                  class="checkbox-outer"
                >
                  <div
                    class="custom-checkbox remembered-false"
                  />
                </div>
                <p
                  class="remembered-label"
                >
                  Stay signed in on this browser?
                </p>
              </div>
              <div
                class="form-error hiding-error"
              />
              <button
                class="form-button"
                type="submit"
              >
                Login
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
